fairlearn/postprocessing/_threshold_optimizer.py

```diff
@@ -38,12 +39,131 @@
+SUPPORTED_UTILITIES = {
+    'selection_rate',
+    'demographic_parity',
+    'false_positive_rate',
+    'false_negative_rate',
+    'true_positive_rate',
+    'true_negative_rate',
+}
+
+# For equalized odds, we only allow objectives that are non-decreasing in true_positives,
+# holding n, positives, negatives, true_negatives, and false_positives fixed.
+SUPPORTED_OBJECTIVES_FOR_EQUALIZED_ODDS = {
+    'accuracy_score',
+}
+
+SUPPORTED_CONSTRAINTS = [*SUPPORTED_UTILITIES, EQUALIZED_ODDS]
+NOT_SUPPORTED_CONSTRAINTS_ERROR_MESSAGE = (
+    "Currently only the following constraints are supported: {}.".format(
+        ", ".join(SUPPORTED_CONSTRAINTS)))
+NOT_SUPPORTED_OBJECTIVES_ERROR_MESSAGE = (
+    "Currently only the following objectives are supported: {}.".format(
+    ", ".join(METRIC_DICT.keys())))
+NOT_SUPPORTED_OBJECTIVES_FOR_EQUALIZED_ODDS_ERROR_MESSAGE = (
+    "Currently only the following objectives are supported: {}.".format(
+    ", ".join(SUPPORTED_OBJECTIVES_FOR_EQUALIZED_ODDS)))
+
+class InterpolatedThresholder(BaseEstimator, MetaEstimatorMixin):
+    """Binary predictor that thresholds continuous predictions of a base estimator.
+
+    At prediction time, the predictor takes as input both standard and sensitive features.
+    Based on the values of sensitive features, it then applies a randomized thresholding
+    transformation according to the provided `interpolation_dict`.
+    
+    :param estimator: base estimator
+    :param dict interpolation_dict: maps sensitive feature values to `Bunch` that describes the
+        interpolation transformation via the following fields:
+        - p0, operation0: with probability p0, operation0 is executed
+        - p1, operation1: with probability p1, operation1 is executed
+        - p_ignore, prediction_constant: two optional fields; if present then the result of
+          the draw of operation0 or operation1 is kept with probability 1 - p_ignore, and gets
+          replaced by prediction_constant with probability p_ignore.
+        The numbers p0 and p1 must be non-negative and add up to 1, operation0 and
+        operation1 must be instances of :class:`ThresholdOperation`, and p_ignore must be
+        between 0 and 1.
+    :param bool prefit: if `True` then the base estimator is not fitted in :meth:`fit`.
+    """
+
+    def __init__(self, estimator, interpolation_dict, prefit=False):
+        self.estimator = estimator
+        self.interpolation_dict = interpolation_dict
+        self.prefit = prefit
+
+    def fit(self, X, y, **kwargs):
+        """Fit the estimator.
+
+        If `prefit` is set to `True` then the base estimator is kept as is.
+        Otherwise it is fitted from the provided arguments.
+        """
+        if self.estimator is None:
+            raise ValueError(ESTIMATOR_ERROR_MESSAGE)
+        
+        if not self.prefit:
+            self.estimator_ = clone(self.estimator).fit(X, y, **kwargs)
+        else:
+            self.estimator_ = self.estimator
```

Comment:
this should also check if the estimator is already fitted. Otherwise it should refit anyway.

Reply:
This actually points out to a deeper issue.

I tried to do it the way you suggested first, but running with `LGBMClassifier` was triggering refitting even though it was already fitted. Obviously they are not implementing sklearn spec correctly, but that might end up being quite common when people create simple wrappers around their non-sklearn classifiers to pull into our library. I am considering two options here:
- Don't make `InterpolatedThresholder` a sklearn estimator; we would not bring in the mixins and would only implement `predict`, but not `fit`. (Alas, I would use the pretty printing of `BaseEstimator.__repr__`
- Throw a warning or even an exception when there's a mismatch between `prefit` and `check_fitted` (and required a manual override for an exception, e.g., by setting an additional argument? or is that too much?)
Thoughts?

Reply:
I dislike not conforming to sklearn because other packages don't follow conventions properly.
I also dislike adding yet another argument, because that would just completely confuse me as the user.

Let me see whether I can write up expected behavior (IMO)
- `prefit=True` and `check_fitted` returns True: no additional fitting
- `prefit=True` and `check_fitted` returns False: user tells us explicitly that it's fitted, so don't fit again, but log warning and if it fails on `predict` provide an explanation. This seems like a typical scenario of "user error" to me. If a user provides `prefit=True` we should never call `fit`.
- `prefit=False` and `check_fitted` returns True: call `fit` anyway because the user seems to want us to refit it.
- `prefit=False` and `check_fitted` returns False: call `fit`

What do you think?

Reply:
The issue with `prefit=True` and warning or raising is that `clone` will keep `prefit=True`, but the underlying estimator will be a fresh unfitted one. To me it's a straightforward issue, `LGBMClassifier` has an issue which can very easily be fixed, so I'd open an issue on https://github.com/microsoft/lightgbm and most probably have it fixed. It'll probably be better received if the issue report comes from withing Microsoft (gut feeling, not sure about your politics ;) )

Reply:
That helps with LightGBM, but that won't fix the same issue for other packages.

This is what I filed with them: https://github.com/microsoft/LightGBM/issues/3014

Reply:
Well, if you go with a nice sklearn compat, then it makes sense for the user to work in that ecosystem I'd say.

Reply:
I kinda want to have a cake and eat it too. @romanlutz : your proposal exactly matches mine.

A slight elaboration:
- if `prefit=True` and `check_fitted=True`: no additional fitting
- if `prefit=True` and `check_fitted=True`: no additional fitting, but give a warning--the warning should say that if the user is running crossvalidation etc., they should do so with `prefit=False` and if the estimator is already fitted, they should file the issue to make it sklearn compliant. 
- if `prefit=False`: always call `fit`

@adrinjalali :  I'm also a bit worried about a silent failure (where the user is setting `prefit=True`, but is unaware that the base estimator is being refitted). This proposal addresses that issue too.

Reply:
I was checking `SelectFromModel` in sklearn, and apparently there we even raise if the user calls `fit` and `prefit=True`, which means users can't give an estimator with `prefit=True` to a `GridSearchCV` for instance. So a warning, or even raising, would be consistent with that.

Reply:
Great, I'll then implement my last suggestion.